### PR TITLE
REFACTOR cookbook for covering new API JSON representation. Part I

### PR DESCRIPTION
**Pull request by jmcanterafonseca**

### CODE REVIEW

**Comment by jmcanterafonseca** [comment]
This is a series of PR trying to align the spec with the Cook book. It has covered list and query operations. Please note that a refactoring of the grouping of operations that work with attributes is expected as well.

File: doc/apiary/v2/fiware-ngsiv2-cookbook.apib [hunk 1]
```diff
@@ -124,43 +206,79 @@ This resource allows you to obtain a list of subresources published by the API:
...
+              "metadata": { } 
             }
          }
         ]
 
-### List all entities which id matches a pattern [GET /v2/entities{?idPattern}]
+### By pattern matching [GET /v2/entities{?idPattern,options}]
 
 + Parameters
     + idPattern: DC_* (optional, string) - Only retrieve entities which id matches this pattern.
```
Comment (kzangeli) [inline doc/apiary/v2/fiware-ngsiv2-cookbook.apib]: DC_*  => DC_.*  ?
Only retrieve entities whose id match this pattern (which => whose)
Now, is it 'match' or 'matches'?
I am not sure, I would avoid the problem by rephrasing:

Only retrieve entities having a matching id.

File: doc/apiary/v2/fiware-ngsiv2-cookbook.apib [hunk 2]
```diff
@@ -36,9 +38,42 @@ This resource allows you to obtain a list of subresources published by the API:
             "registrations_url": "/v2/registrations"
         }
 
-## Listing context entities [/v2/entities{?type,id,idPattern,attrs,limit,offset,options}]
+# Group Entity Collection
 
-### List all entities [GET /v2/entities]
+## List all entities [/v2/entities]
+
+### Normalized mode [GET /v2/entities]
```
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-cookbook.apib]: "Normalized", "Cannonical", etc. terms are obsolete.

What about just "Default mode" ?

File: doc/apiary/v2/fiware-ngsiv2-cookbook.apib [hunk 3]
```diff
@@ -49,29 +84,62 @@ This resource allows you to obtain a list of subresources published by the API:
...
+    + options: count
+    + limit: 2
+    + offset: 2
+
++ Response 200 (application/json)
+
+    + Headers
+
+            X-Total-Count: 30
+            Link: http://context.example.org/v2/entities?offset=4&limit=2; rel="next", http://context.example.org/v2/entities?offset=28&limit=2; rel="last", http://context.example.org/v2/entities?offset=0&limit=2; rel="first", http://context.example.org/v2/entities?offset=2&limit=2; rel="prev"
```
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-cookbook.apib]: Not sura how Apiary/FABRE will render this, but at least in the Github diff view it is quite long...

File: doc/apiary/v2/fiware-ngsiv2-cookbook.apib [hunk 4]
```diff
@@ -172,52 +290,125 @@ This resource allows you to obtain a list of subresources published by the API:
...
+
++ Parameters
+    + entityId: Bcn_Welt (required, string) - Entity ID
+    + attrs: mistake (optional, string) - Attribute to be retrieved
+
++ Response 200 (application/json)
+
+        {
+           "type": "Room",
+            "id": "Boe_Idearium"
```
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-cookbook.apib]: Missplaced indent (only one whitespace)

File: doc/apiary/v2/fiware-ngsiv2-cookbook.apib [hunk 5]
```diff
@@ -172,52 +290,125 @@ This resource allows you to obtain a list of subresources published by the API:
...
     The number of values will depend on the 'geometry' attribute.
-    + geometry: circle;radius=10000 (optional, string) - It can be any type defined by the
-    GeoJSON spec or 'circle;radius=&lt;a radius&gt;'.
+    + geometry: circle;radius=10000 (optional, string) - Geospatial area for the query. 
     The 'exterior' modifier allows to indicate that the referred area is the outside
     of the region, for instance, 'polygon;exterior'
     + limit: 20 - Limit the number of results (optional, number) - Maximum number of results to return
     + attrs: seatsNumber (optional, string) - Attributes to be returned
-    + options: count (optional, string) - count indicates that the total number of records must be returned
+    + options: count,keyValues (optional, string) - count indicates that the total number of records must be returned
```
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-cookbook.apib]: keyValues has been added to the options list, but the description stills referes only to count. Maybe a neutral description wuold be better (i.e. describing options itself, without referring to count or keyValues).

**Comment by fgalan** [comment]
Some minor comments in the PR... Good to see progress in the cookbook! Many users of the beta functionality for NGSIv2 are looking at it, so we should have it as synchronized as possible with reference .apib (and with Orion implementation itself).

File: doc/apiary/v2/fiware-ngsiv2-cookbook.apib (hunk 1 above)
Comment (jmcanterafonseca) [inline doc/apiary/v2/fiware-ngsiv2-cookbook.apib]: fixed https://github.com/jmcanterafonseca/fiware-orion/commit/82cb41f6c5be3650f5560f7728973f3331cc484b

File: doc/apiary/v2/fiware-ngsiv2-cookbook.apib (hunk 2 above)
Comment (jmcanterafonseca) [inline doc/apiary/v2/fiware-ngsiv2-cookbook.apib]: fixed https://github.com/jmcanterafonseca/fiware-orion/commit/82cb41f6c5be3650f5560f7728973f3331cc484b

File: doc/apiary/v2/fiware-ngsiv2-cookbook.apib (hunk 3 above)
Comment (jmcanterafonseca) [inline doc/apiary/v2/fiware-ngsiv2-cookbook.apib]: fixed https://github.com/jmcanterafonseca/fiware-orion/commit/82cb41f6c5be3650f5560f7728973f3331cc484b

File: doc/apiary/v2/fiware-ngsiv2-cookbook.apib (hunk 4 above)
Comment (jmcanterafonseca) [inline doc/apiary/v2/fiware-ngsiv2-cookbook.apib]: fixed https://github.com/jmcanterafonseca/fiware-orion/commit/82cb41f6c5be3650f5560f7728973f3331cc484b

File: doc/apiary/v2/fiware-ngsiv2-cookbook.apib (hunk 5 above)
Comment (jmcanterafonseca) [inline doc/apiary/v2/fiware-ngsiv2-cookbook.apib]: fixed https://github.com/jmcanterafonseca/fiware-orion/commit/82cb41f6c5be3650f5560f7728973f3331cc484b

**Comment by fgalan** [comment]
LGTM

Note that LGTM from other participants in the review ( @kzangeli in particular) is neded before this PR may be merged.

**Comment by kzangeli** [comment]
LGTM

**Pull request merged**